### PR TITLE
Performance improvement

### PR DESCRIPTION
**Pull request by asimarslan**
performance improvement on serialization bits, codecs and lazyDictioary introduced

### CODE REVIEW

**Comment by asimarslan** [comment]
verify

File: Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapNearCacheProxy.cs [hunk 1]
```diff
@@ -256,45 +258,46 @@ protected override object ExecuteOnKeyInternal(IData keyData, IEntryProcessor en
...
+                for (int i = 0; i < keyList.Count; i++)
                 {
-                    list.Remove(keyData);
-                    resultingKeyValuePairs.Add(keyData);
-                    resultingKeyValuePairs.Add(value);
+                    var keyData = (IData)keyList[i];
+                    object value;
+                    if (_nearCache.TryGetValue(keyData, out value))
+                    {
+                        keyList.Remove(keyData);
```
Comment (sancar) [inline Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapNearCacheProxy.cs]: We are iterating over the keyList with indexes and calling remove in the middle. It looks like we will have array index out of bound exception.

Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapNearCacheProxy.cs]: reverted the iteration

File: Hazelcast.Net/Hazelcast.Client.Protocol.Codec/MapGetAllCodec.cs [hunk 2]
```diff
@@ -22,53 +25,45 @@ namespace Hazelcast.Client.Protocol.Codec
...
     internal static class MapGetAllCodec
     {
-        private static int CalculateRequestDataSize(string name, IList<IData> keys)
+        private static int CalculateRequestDataSize(string name, ArrayList keys)
         {
             var dataSize = ClientMessage.HeaderSize;
             dataSize += ParameterUtil.CalculateDataSize(name);
             dataSize += Bits.IntSizeInBytes;
-            foreach (var keysItem in keys)
+            for (int i = 0; i < keys.Count; i++)
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Protocol.Codec/MapGetAllCodec.cs]: on .Net foreach is slower than for loop

File: Hazelcast.Net/Hazelcast.Client.Protocol.Util/BufferBuilder.cs [hunk 3]
```diff
@@ -100,7 +100,7 @@ private void EnsureCapacity(int additionalCapacity)
             {
                 var newCapacity = QuickMath.NextPowerOfTwo(requiredCapacity);
                 var newBuffer = new byte[newCapacity];
-                Array.Copy(_protocolBuffer.ByteArray(), 0, newBuffer, 0, _capacity);
+                Buffer.BlockCopy(_protocolBuffer.ByteArray(), 0, newBuffer, 0, _capacity);
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Protocol.Util/BufferBuilder.cs]: Buffer.BlockCopy is faster than Array.Copy. both are native calls and I'm not sure of the reason. I can measure improvement.

File: Hazelcast.Net/Hazelcast.Client.Protocol.Util/SafeBuffer.cs [hunk 4]
```diff
@@ -25,52 +25,49 @@ namespace Hazelcast.Client.Protocol.Util
...
-        private const bool ShouldBoundsCheck = true;
         private byte[] _byteArray;
 
         public SafeBuffer(byte[] buffer)
         {
             Wrap(buffer);
         }
 
-        public virtual void PutLong(int index, long value)
+        public  void PutLong(int index, long value)
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Protocol.Util/SafeBuffer.cs]: yes virtual methods and interface calls adds extra overhead.

File: Hazelcast.Net/Hazelcast.Client.Protocol.Util/SafeBuffer.cs [hunk 5]
```diff
@@ -92,78 +89,46 @@ public byte[] ByteArray()
...
-        public virtual void GetBytes(int index, byte[] dst)
+        public  void GetBytes(int index, byte[] dst)
         {
             GetBytes(index, dst, 0, dst.Length);
         }
 
-        public virtual void GetBytes(int index, byte[] dst, int offset, int length)
+        public  void GetBytes(int index, byte[] dst, int offset, int length)
         {
-            BoundsCheck(index, length);
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Protocol.Util/SafeBuffer.cs]: I removed it as it's unnecessary. This is bound check is on all of the hot path so every peny count on this kink of hot paths.

File: Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapProxy.cs [hunk 6]
```diff
@@ -769,36 +754,39 @@ public bool IsEmpty()
         public void PutAll(IDictionary<TKey, TValue> m)
         {
             var partitionService = GetContext().GetPartitionService();
-            var partitions = new Dictionary<int, IDictionary<IData, IData>>(partitionService.GetPartitionCount());
-            foreach (var kvp in m)
+            var partitionCount = partitionService.GetPartitionCount();
+            var partitions = new ArrayList(partitionCount);
+            for (var i = 0; i < partitionCount; i++)
+            {
+                partitions.Add(ArrayList.Synchronized(new ArrayList()));
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapProxy.cs]: yes the next parallel loop will use it but when I benchmark it, I saw itS even faster than concurrentQueue as most of the time is used on the serialization and threads rarely make sync calls.

File: Hazelcast.Net/Hazelcast.Util/ReadOnlyLazyList.cs [hunk 7]
```diff
@@ -43,18 +44,31 @@ IEnumerator IEnumerable.GetEnumerator()
...
+            foreach (var dValue in list)
+            {
+                if (dValue is IData)
+                {
+                    if (itemData.Equals(dValue))
+                        return true;
+                }
+                else
+                {
+                    if (itemData.Equals(serializationService.ToData(dValue)))
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Util/ReadOnlyLazyList.cs]: A second call to Contains is quite a rare case for this list. 99% will just loop over it. But if I cache it, that will be a memory overhead for every case.
Moreover contains method is just implemented just not to leave it empty.

File: Hazelcast.Net/Hazelcast.Client.Protocol.Codec/MapEntriesWithPredicateCodec.cs [hunk 8]
```diff
@@ -41,25 +42,16 @@ internal static ClientMessage EncodeRequest(string name, IData predicate)
             return clientMessage;
         }
 
-        internal class ResponseParameters
+        internal static void DecodeResponse(IClientMessage clientMessage, ConcurrentQueue<KeyValuePair<IData, object>> result)
```
[the remark this answers is not in the document]
Comment (asimarslan) [reply Hazelcast.Net/Hazelcast.Client.Protocol.Codec/MapEntriesWithPredicateCodec.cs]: This is a IMap optimization that we can backport in 4.0 series to all codecs

**Comment by asimarslan** [comment]
verify

**Comment by asimarslan** [comment]
run-net

**Pull request merged**